encoding/multi_reader_iterator.go

```diff
@@ -125,54 +124,31 @@ func (it *multiReaderIterator) moveToNext() {
 		}
 	}
 
-	if it.iters.Len() == 0 && !it.hasError() {
+	if it.iters.len() == 0 && !it.hasError() {
 		// No iterators were added, move to next
 		it.moveToNext()
 	}
 }
 
 func (it *multiReaderIterator) moveIteratorsToNext() {
-	iter := heap.Pop(&it.iters).(ReaderIterator)
-	prev, _, _ := iter.Current()
-
-	if it.moveIteratorToValidNext(iter) {
-		heap.Push(&it.iters, iter)
-	} else {
-		iter.Close()
-	}
-
-	if it.iters.Len() == 0 {
-		return
-	}
-
-	curr, _, _ := it.Current()
-	if curr.Timestamp.Equal(prev.Timestamp) {
-		// Dedupe
-		it.moveIteratorsToNext()
-	}
-}
+	for {
+		prev := it.iters.at()
+		next, err := it.iters.moveToValidNext()
+		if err != nil {
```

Comment:
So it's definitely not possible to get here without `it.err` being nil as `moveIteratorsToNext` is only called from `moveToNext` which is only called after checking `hasNext` which checks `!it.hasError()`.

I'll add a check anyhow though.